solver/src/main/java/org/chocosolver/solver/search/strategy/selectors/variables/ConflictHistorySearch.java

```diff
@@ -0,0 +1,133 @@
+
+import gnu.trove.map.TObjectDoubleMap;
+import gnu.trove.map.TObjectIntMap;
+import gnu.trove.map.hash.TObjectDoubleHashMap;
+import gnu.trove.map.hash.TObjectIntHashMap;
+import org.chocosolver.solver.Solver;
+import org.chocosolver.solver.constraints.Propagator;
+import org.chocosolver.solver.exception.ContradictionException;
+import org.chocosolver.solver.search.loop.monitors.IMonitorContradiction;
+import org.chocosolver.solver.search.loop.monitors.IMonitorRestart;
+import org.chocosolver.solver.search.strategy.selectors.values.IntValueSelector;
+import org.chocosolver.solver.variables.IntVar;
+
+/**
+ * Source: "Conflist History Based Branching Heuristic for CSP Solving", Habet and Terrioux.
+ * <p>
+ * Project: choco.
+ *
+ * @author Charles Prud'homme
+ * @since 25/02/2020.
+ */
+@SuppressWarnings("rawtypes")
+public class ConflictHistorySearch
+    extends AbsctractCriterionBasedStrategy
+    implements IMonitorContradiction, IMonitorRestart {
+
+    /**
+     * Decreasing step for {@link #a}.
+     */
+    private static final double STEP = 10e-6;
+    private static final double D = 10e-4;
+    private static final double DECAY = .995;
+    /**
+     * Score of each propagator.
+     */
+    private final TObjectDoubleMap<Propagator> q = new TObjectDoubleHashMap<>(10, 0.5f, 0.0);
+    /**
+     * Step-size, 0 < a < 1.
+     */
+    private double a = .4d;
+    /**
+     * Reward value of each propagator.
+     */
+    private final TObjectDoubleMap<Propagator> r = new TObjectDoubleHashMap<>(10, 0.5f, 0.0);
+    /**
+     * The number of conflicts which have occurred since the beginning of the search.
+     */
+    private int conflicts = 0;
+    /**
+     * Last {@link #conflicts} value where a propagator led to a failure.
+     */
+    private final TObjectIntMap<Propagator> conflict = new TObjectIntHashMap<>(10, 0.5f, 0);
+
+
+    public ConflictHistorySearch(IntVar[] vars, long seed, IntValueSelector valueSelector) {
+        super(vars, seed, valueSelector);
+        vars[0].getModel().getSolver().plugMonitor(this);
+    }
+
+    @Override
+    public boolean init() {
+        Solver solver = vars[0].getModel().getSolver();
+        if(!solver.getSearchMonitors().contains(this)) {
+            vars[0].getModel().getSolver().plugMonitor(this);
+        }
+        return true;
+    }
+
+    @Override
+    public void remove() {
+        Solver solver = vars[0].getModel().getSolver();
+        if(solver.getSearchMonitors().contains(this)) {
+            vars[0].getModel().getSolver().unplugMonitor(this);
+        }
+    }
+
+
+    @Override
+    public void onContradiction(ContradictionException cex) {
+        if (cex.c instanceof Propagator) {
+            Propagator p = (Propagator) cex.c;
+            double qj = q.get(p);
+            // update q
+            q.put(p, (1 - a) * qj + a * r.get(p));
+            // update r
+            double rj = r.get(p);
+            r.put(p, 1d / (conflicts - rj + 1));
```

Comment:
I do not agree with the update of the reward. According to the paper, the reward used inside the update of q(cj) uses the reward (1d / (conflicts - rj + 1)) whereas rj is simply the latest value of variable conflicts when this specific constraint (here Propagator) fails.

So, for me, the code should be : `r.put(p, 1d / (conflicts - conflict.get(p) + 1));`